### PR TITLE
Revamps autolinker to fix bugs and improve efficiency

### LINKED ISSUE #6497: RC3, URL parsing oddity
#### Description
Originally opened a topic about this at https://www.simplemachines.org/community/index.php?topic=576638 
This works correctly on 2.0, and is linked automatically

https://puheenvuoro.uusisuomi.fi/jukka-savorinen/olen-selvittanyt-miten-maailmankaikkeus-toimii-%f0%9f%98%83/

On my RC test install though, for some reason it doesn't. It does work, if I manually add URL tags.
I don't really know what else to say, except that I could assume this is something to do with the encoded emoji in the URL.

### LINKED ISSUE #6685: the message body was left empty
#### Description
created topic with subject test and body https://en.wikipedia.org/wiki/Kotlin_(programming_language)
clicked on post and was greeted with the message "The message body was left empty."

### Steps to reproduce
1. create new topic, subject test, body https://en.wikipedia.org/wiki/Kotlin_(programming_language)
2. click post see message "The message body was left empty."

### Environment (complete as necessary)
- Version/Git revision: as of 12am EDT last night
- Database Type: mariadb
- Database Version: 10.4.10
- PHP Version: 7.3.12

### Additional information/references

### PR DESCRIPTION
**Pull request by Sesquipedalian**
Fixes #6685
Closes #6691
Closes #6692
Fixes #6497

This uses some more advanced PCRE techniques that I have learned since the last time I revamped the autolinker. Not only does this solve bugs (particularly the aforementioned showstopper bug involving parentheses in an URI), but it also improves efficiency by making more extensive use of PCRE subroutines. Additionally, this takes advantage of PCRE conditional logic to test for specific conditions based on the URI scheme, allowing us to do better at avoiding false positives. Along the way, I also found and fixed a couple of edge case bugs in my `validate_iri()`, `sanitize_iri()`, `iri_to_url()`, and `url_to_iri()` functions.

### CODE REVIEW

**Comment by Sesquipedalian** [comment]
In the comments on #6691, @BrickOzp asked if anyone had anything to use for a unit test for the autolinker. Here is the set of IRI strings I used during testing:
```
example.biz
www.example.info
ftp://user:password@ftp.example.com
ftp://ftp.is.co.za/rfc/rfc1808.txt
gopher://spinaltap.micro.umn.edu/00/Weather/California/Los%20Angeles
http://www.math.uio.no/faq/compression-faq/part1.html
mailto:mduerst@ifi.unizh.ch
news:comp.infosystems.www.servers.unix
telnet://melvyl.ucop.edu/
foobar//www.example.com
http://høns.dk
http://ar.wikipedia.org/wiki/النطاقات_العربية
http://example.com/foo_(bar_baz)
http://example.com/foo_(bar_baz
(http://example.com/foo_bar_baz)
http://example.com/(bar_(qux)_baz)_(foo)?test=test;foo=bar
example.com/?test=[(foo)]{bar}
http://www.example.com/a/test?one=for#{all”
http://www.example.com/a/test?one=for<all>#derp
mailto:jsmith@example.com?subject=A%20Test&body=My%20idea%20is%3A%20%0A
file:///Users/foo/bar.txt
http://example.com/
http://example.com//
http://example.com/foo.php?
foo://bar
facetime://+19995551234
Have you been to example.com? It is exemplary!
Have you been to example.com?It is exemplary!
Have you been to //example.com?It is exemplary!
Have you been to http://example.com?It is exemplary!
http://123.4.56.78/
https://[2001:db8:85a3:8d3:1319:8a2e:370:7348]:443/foo.html
http://[::1]/foo.html
//foo.asghetyh
//foo.com
qwer:111
ssh://localhost
ssh://toaster.local
ssh:111
mailto:mduérst@ifi.unizh.ch
jsmith@example.com
For an example of a valuably variable foo, go to http://www.example.com/index.php?var[foo]=value!
Have you been to http://www.example.com/index.php?var[foo]=value?
In parentheses (http://www.example.com/index.php?var[foo]=value), will this still be autolinked correctly? 
What about in brackets [http://www.example.com/index.php?var[foo]=value]? 
```

**Comment by Sesquipedalian** [comment]
In the comments on #6691, @BrickOzp asked if anyone had anything to use for a unit test for the autolinker. Here is the set of IRI strings I used during testing:
```
example.biz
www.example.info
ftp://user:password@ftp.example.com
ftp://ftp.is.co.za/rfc/rfc1808.txt
gopher://spinaltap.micro.umn.edu/00/Weather/California/Los%20Angeles
http://www.math.uio.no/faq/compression-faq/part1.html
mailto:mduerst@ifi.unizh.ch
news:comp.infosystems.www.servers.unix
telnet://melvyl.ucop.edu/
foobar//www.example.com
http://høns.dk
http://ar.wikipedia.org/wiki/النطاقات_العربية
http://example.com/foo_(bar_baz)
http://example.com/foo_(bar_baz
(http://example.com/foo_bar_baz)
http://example.com/(bar_(qux)_baz)_(foo)?test=test;foo=bar
example.com/?test=[(foo)]{bar}
http://www.example.com/a/test?one=for#{all”
http://www.example.com/a/test?one=for<all>#derp
mailto:jsmith@example.com?subject=A%20Test&body=My%20idea%20is%3A%20%0A
file:///Users/foo/bar.txt
http://example.com/
http://example.com//
http://example.com/foo.php?
foo://bar
facetime://+19995551234
Have you been to example.com? It is exemplary!
Have you been to example.com?It is exemplary!
Have you been to //example.com?It is exemplary!
Have you been to http://example.com?It is exemplary!
http://123.4.56.78/
https://[2001:db8:85a3:8d3:1319:8a2e:370:7348]:443/foo.html
http://[::1]/foo.html
//foo.asghetyh
//foo.com
qwer:111
ssh://localhost
ssh://toaster.local
ssh:111
mailto:mduérst@ifi.unizh.ch
jsmith@example.com
For an example of a valuably variable foo, go to http://www.example.com/index.php?var[foo]=value!
Have you been to http://www.example.com/index.php?var[foo]=value?
In parentheses (http://www.example.com/index.php?var[foo]=value), will this still be autolinked correctly? 
What about in brackets [http://www.example.com/index.php?var[foo]=value]? 
```

The expected HTML output is:
```
<a href="//example.biz" class="bbc_link" target="_blank" rel="noopener">example.biz</a><br><a href="//www.example.info" class="bbc_link" target="_blank" rel="noopener">www.example.info</a><br><a href="ftp://user:password@ftp.example.com" class="bbc_link" target="_blank" rel="noopener">ftp://user:password@ftp.example.com</a><br><a href="ftp://ftp.is.co.za/rfc/rfc1808.txt" class="bbc_link" target="_blank" rel="noopener">ftp://ftp.is.co.za/rfc/rfc1808.txt</a><br><a href="gopher://spinaltap.micro.umn.edu/00/Weather/California/Los%20Angeles" class="bbc_link" target="_blank" rel="noopener">gopher://spinaltap.micro.umn.edu/00/Weather/California/Los%20Angeles</a><br><a href="http://www.math.uio.no/faq/compression-faq/part1.html" class="bbc_link" target="_blank" rel="noopener">http://www.math.uio.no/faq/compression-faq/part1.html</a><br><a href="mailto:mduerst@ifi.unizh.ch" class="bbc_email">mailto:mduerst@ifi.unizh.ch</a><br><a href="news:comp.infosystems.www.servers.unix" class="bbc_link" target="_blank" rel="noopener">news:comp.infosystems.www.servers.unix</a><br><a href="telnet://melvyl.ucop.edu/" class="bbc_link" target="_blank" rel="noopener">telnet://melvyl.ucop.edu/</a><br>foobar<a href="//www.example.com" class="bbc_link" target="_blank" rel="noopener">//www.example.com</a><br><a href="http://høns.dk" class="bbc_link" target="_blank" rel="noopener">http://høns.dk</a><br><a href="http://ar.wikipedia.org/wiki/النطاقات_العربية" class="bbc_link" target="_blank" rel="noopener">http://ar.wikipedia.org/wiki/النطاقات_العربية</a><br><a href="http://example.com/foo_(bar_baz)" class="bbc_link" target="_blank" rel="noopener">http://example.com/foo_(bar_baz)</a><br><a href="http://example.com/foo_(bar_baz" class="bbc_link" target="_blank" rel="noopener">http://example.com/foo_(bar_baz</a><br>(<a href="http://example.com/foo_bar_baz" class="bbc_link" target="_blank" rel="noopener">http://example.com/foo_bar_baz</a>)<br><a href="http://example.com/(bar_(qux)_baz)_(foo)?test=test;foo=bar" class="bbc_link" target="_blank" rel="noopener">http://example.com/(bar_(qux)_baz)_(foo)?test=test;foo=bar</a><br><a href="//example.com/?test=[(foo)]{bar}" class="bbc_link" target="_blank" rel="noopener">example.com/?test=[(foo)]{bar}</a><br><a href="http://www.example.com/a/test?one=for#{all" class="bbc_link" target="_blank" rel="noopener">http://www.example.com/a/test?one=for#{all</a>"<br><a href="http://www.example.com/a/test?one=for" class="bbc_link" target="_blank" rel="noopener">http://www.example.com/a/test?one=for</a>&lt;all&gt;#derp<br><a href="mailto:jsmith@example.com?subject=A%20Test&amp;body=My%20idea%20is%3A%20%0A" class="bbc_email">mailto:jsmith@example.com?subject=A%20Test&amp;body=My%20idea%20is%3A%20%0A</a><br><a href="file:///Users/foo/bar.txt" class="bbc_link" target="_blank" rel="noopener">file:///Users/foo/bar.txt</a><br><a href="http://example.com/" class="bbc_link" target="_blank" rel="noopener">http://example.com/</a><br><a href="http://example.com/" class="bbc_link" target="_blank" rel="noopener">http://example.com/</a>/<br><a href="http://example.com/foo.php" class="bbc_link" target="_blank" rel="noopener">http://example.com/foo.php</a>?<br><a href="foo://bar" class="bbc_link" target="_blank" rel="noopener">foo://bar</a><br><a href="facetime://+19995551234" class="bbc_link" target="_blank" rel="noopener">facetime://+19995551234</a><br>Have you been to <a href="//example.com" class="bbc_link" target="_blank" rel="noopener">example.com</a>? It is exemplary!<br>Have you been to <a href="//example.com" class="bbc_link" target="_blank" rel="noopener">example.com</a>?It is exemplary!<br>Have you been to <a href="//example.com" class="bbc_link" target="_blank" rel="noopener">//example.com</a>?It is exemplary!<br>Have you been to <a href="http://example.com" class="bbc_link" target="_blank" rel="noopener">http://example.com</a>?It is exemplary!<br><a href="http://123.4.56.78/" class="bbc_link" target="_blank" rel="noopener">http://123.4.56.78/</a><br><a href="https://[2001:db8:85a3:8d3:1319:8a2e:370:7348]:443/foo.html" class="bbc_link" target="_blank" rel="noopener">https://[2001:db8:85a3:8d3:1319:8a2e:370:7348]:443/foo.html</a><br><a href="http://[::1]/foo.html" class="bbc_link" target="_blank" rel="noopener">http://[::1]/foo.html</a><br>//foo.asghetyh<br><a href="//foo.com" class="bbc_link" target="_blank" rel="noopener">//foo.com</a><br>qwer:111<br><a href="ssh://localhost" class="bbc_link" target="_blank" rel="noopener">ssh://localhost</a><br><a href="ssh://toaster.local" class="bbc_link" target="_blank" rel="noopener">ssh://toaster.local</a><br>ssh:111<br><a href="mailto:mduérst@ifi.unizh.ch" class="bbc_email">mailto:mduérst@ifi.unizh.ch</a><br><a href="mailto:jsmith@example.com" class="bbc_email">jsmith@example.com</a><br>For an example of a valuably variable foo, go to <a href="http://www.example.com/index.php?var[foo]=value" class="bbc_link" target="_blank" rel="noopener">http://www.example.com/index.php?var[foo]=value</a>!<br>Have you been to <a href="http://www.example.com/index.php?var[foo]=value" class="bbc_link" target="_blank" rel="noopener">http://www.example.com/index.php?var[foo]=value</a>?<br>In parentheses (<a href="http://www.example.com/index.php?var[foo]=value" class="bbc_link" target="_blank" rel="noopener">http://www.example.com/index.php?var[foo]=value</a>), will this still be autolinked correctly? <br>What about in brackets [<a href="http://www.example.com/index.php?var[foo]=value" class="bbc_link" target="_blank" rel="noopener">http://www.example.com/index.php?var[foo]=value</a>]? <br>
```
Screen Shot 2021-05-22 at 4 28 25 PM

**Review by BrickOzp — COMMENTED** [review]
FILTER_FLAG_EMAIL_UNICODE was introduced in PHP 7.1.0
With PHP 5.6 it will give this error in the error log Use of undefined constant FILTER_FLAG_EMAIL_UNICODE - assumed 'FILTER_FLAG_EMAIL_UNICODE'

I havn't found any workaround, but silencing the warning seems to remove the error. @filter_var...
mailto:mduérst@ifi.unizh.ch will not be linked though

**Comment by BrickOzp** [comment]
The ipv6 addresses does not get autolinked with php 5.6 either
https://[2001:db8:85a3:8d3:1319:8a2e:370:7348]:443/foo.html
http://[::1]/foo.html

**Comment by BrickOzp** [comment]
This fixes https://github.com/SimpleMachines/SMF2.1/issues/6497 as well

**Comment by Sesquipedalian** [comment]
Thanks, @BrickOzp. Should be fixed now.

File: Sources/Subs.php [hunk 1]
```diff
@@ -7393,6 +7393,15 @@ function validate_iri($iri, $flags = null)
 {
 	$url = iri_to_url($iri);
 
+	// PHP 5 doesn't recognize IPv6 addresses in the URL host.
+	if (version_compare(phpversion(), '7.0.0', '<'))
+	{
+		$host = parse_url((strpos($url, '//') === 0 ? 'http:' : '') . $url, PHP_URL_HOST);
+
+		if (strpos($host, '[') === 0 && strpos($host, ']') === strlen($host) && strpos($host, ':') !== false)
```
Comment (BrickOzp) [inline Sources/Subs.php]: Cool workaround but should be strpos($host, ']') === strlen($host) **- 1** to work

**Comment by BrickOzp** [comment]
Mailto with non-ascii chars now works.
See inline comment to get ipv6 to work.

File: Sources/Subs.php (hunk 1 above)
Comment (Sesquipedalian) [reply Sources/Subs.php]: Derp.

**Review by BrickOzp — APPROVED** [review]
